pkg/controllers/dynakube/activegate/internal/tls/reconciler.go

```diff
@@ -0,0 +1,171 @@
+
+	return r.deleteSelfSignedTLSSecret(ctx)
+}
+
+func (r *Reconciler) reconcileSelfSignedTLSSecret(ctx context.Context) error {
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	_, err := query.Get(ctx, types.NamespacedName{
+		Name:      r.dk.ActiveGate().GetTlsSecretName(),
+		Namespace: r.dk.Namespace,
+	})
+
+	if err != nil && k8serrors.IsNotFound(err) {
+		return r.createSelfSignedTLSSecret(ctx)
+	}
+
+	if err != nil {
+		conditions.SetKubeApiError(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) deleteSelfSignedTLSSecret(ctx context.Context) error {
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	return query.Delete(ctx, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.dk.ActiveGate().GetTlsSecretName(),
+			Namespace: r.dk.Namespace,
+		},
+	})
+}
+
+func (r *Reconciler) createSelfSignedTLSSecret(ctx context.Context) error {
+	cert, err := certificates.New(r.timeProvider)
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	cert.Cert.DNSNames = getCertificateAltNames(r.dk.Name)
+	cert.Cert.KeyUsage = x509.KeyUsageKeyEncipherment | x509.KeyUsageDataEncipherment
+	cert.Cert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+	cert.Cert.Subject.CommonName = r.dk.Name + activeGateSelfSignedTLSCommonNameSuffix
+	cert.Cert.IPAddresses = getCertificateAltIPs(r.dk.Status.ActiveGate.ServiceIPs)
+
+	err = cert.SelfSign()
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	pemCert, pemPk, err := cert.ToPEM()
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	coreLabels := k8slabels.NewCoreLabels(r.dk.Name, k8slabels.ActiveGateComponentLabel)
+	secretData := map[string][]byte{
+		consts.TLSCrtDataName: pemCert,
+		consts.TLSKeyDataName: pemPk,
+		tlsCrtDataName:        pemCert,
+	}
+
+	secret, err := k8ssecret.Build(r.dk, r.dk.ActiveGate().GetTlsSecretName(), secretData, k8ssecret.SetLabels(coreLabels.BuildLabels()))
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	secret.Type = corev1.SecretTypeOpaque
+
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	err = query.Create(ctx, secret)
+	if err != nil {
+		conditions.SetKubeApiError(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	conditions.SetSecretCreated(r.dk.Conditions(), conditionType, secret.Name)
+
+	return nil
+}
+
+func getCertificateAltNames(dkName string) []string {
+	return []string{
+		dkName + activeGateSelfSignedTLSCommonNameSuffix,
+		dkName + activeGateSelfSignedTLSCommonNameSuffix + ".svc",
+		dkName + activeGateSelfSignedTLSCommonNameSuffix + ".svc.cluster.local",
+	}
+}
+
+func getCertificateAltIPs(ips []string) []net.IP {
+	altIPs := []net.IP{}
+
+	for _, ip := range ips {
+		altIPs = append(altIPs, net.ParseIP(ip))
```

Comment:
net.ParseIP returns nil if it can't parse it, may be it's better to add small check here?

Reply:
added.
It's a serious error (blocks creation of the cert) because it means a bug in Golang libraries or K8S (the source of the IP string is K8S service). There is no point to create the certificate without IP address because OSAgent needs it.